src/lib/storage/index/abstract_table_index.hpp

```diff
@@ -0,0 +1,194 @@
+#pragma once
+
+#include <memory>
+#include <utility>
+#include <vector>
+
+#include "all_type_variant.hpp"
+#include "table_index_type.hpp"
+#include "types.hpp"
+
+namespace opossum {
+
+/**
+ * Basic forward iterator type for iteration over RowIDs, e.g. for table indexes. The default implementation of the
+ * virtual properties is meant to represent the iterator of an empty collection. Therefore it shall not be dereferenced
+ * (like and end-iterator), increments do not make a change, and equality comparisons with other instances of this class
+ * always result in true.
+ */
+class BaseTableIndexIterator {
+ public:
+  using iterator_category = std::forward_iterator_tag;
+  using value_type = const RowID;
+  using difference_type = std::ptrdiff_t;
+  using pointer = const RowID*;
+  using reference = const RowID&;
+
+  BaseTableIndexIterator(const BaseTableIndexIterator& itr) = default;
+  BaseTableIndexIterator() = default;
+  virtual ~BaseTableIndexIterator() = default;
+
+  virtual reference operator*() const {
+    throw std::logic_error("cannot dereference on empty iterator");
+  }
+  virtual BaseTableIndexIterator& operator++() {
+    return *this;
+  }
+  virtual bool operator==(const BaseTableIndexIterator& other) const {
+    return true;
+  }
+  virtual bool operator!=(const BaseTableIndexIterator& other) const {
+    return false;
+  }
+
+  virtual std::shared_ptr<BaseTableIndexIterator> clone() const {
+    return std::make_shared<BaseTableIndexIterator>();
+  }
+};
+
+/**
+ * Wrapper class that implements an iterator interface and holds a pointer to a BaseTableIndexIterator. This wrapper is
+ * required to allow runtime polymorphism without the need to directly pass pointers to iterators throughout the
+ * codebase. It also provides copy construction and assignment facilities to easily duplicate other IteratorWrappers
+ * including their underlying iterators. This is especially important, because the iterator type is a forward iterator
+ * instead of a random access iterator, so if an iterator instance has to be retained before a manipulating call, e.g.,
+ * when calling it on std::distance, a copy has to be made beforehand.
+ */
+class IteratorWrapper {
```

Comment:
Can we get rid of the two levels of abstractions, i.e., IteratorWrapper and BaseTableIndexIterator?